### PR TITLE
cherry-pick refactor(source): enhance parquet file source (#19221) to branch release-2.1

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
As title, due to dependencies. this actually cherry picked both https://github.com/risingwavelabs/risingwave/pull/19221 and https://github.com/risingwavelabs/risingwave/pull/18967.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/source/filesystem/opendal_source/opendal_reader.rs [hunk 1]
```diff
@@ -215,3 +229,61 @@ impl<Src: OpendalSource> OpendalReader<Src> {
...
+///
+/// # Parameters
+/// - `columns`: A vector of `Column` representing the user's requested schema.
+/// - `metadata`: A reference to `FileMetaData` containing the schema and metadata of the Parquet file.
+///
+/// # Returns
+/// - A `ConnectorResult<Vec<usize>>`, which contains the indices of the valid columns in the
+///   Parquet file schema that match the requested schema. If an error occurs during processing,
+///   it returns an appropriate error.
+pub fn extract_valid_column_indices(
```
Comment (hzxa21) [inline src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:248]: These changes are not in #19221. Are we combining #19221 and #18967 in this cherry pick?

Comment (wcy-fdu) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:248]: Yes, there are dependencies. I actually cherry picked both these two PRs.

Comment (hzxa21) [reply src/connector/src/source/filesystem/opendal_source/opendal_reader.rs:248]: I see. Can you modify the PR description and the title? It is easier to track in this case.

**Review by hzxa21 — APPROVED** [review]
Rest LGTM

**Review by lmatz — APPROVED** [review]

**Pull request merged**